### PR TITLE
PE-2930: Let the "Computing snapshot" modal to be cancellable

### PR DESCRIPTION
**Pull request by matibat**
--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/11t1bdhsbckho
iOS release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:ios:06ea33a95a2e0d42ffce07/releases/2aiq8t9tiad50

### CODE REVIEW

**Review by matibat — COMMENTED** [review]
https://user-images.githubusercontent.com/22564541/217059715-af155b78-759d-48e9-a15b-f211b6dfe939.mov

File: lib/components/create_snapshot_dialog.dart [hunk 1]
```diff
@@ -77,13 +77,29 @@ Widget _loadingDialog(
...
+    progressDescription: Text(
+      snapshotCubitState is ComputingSnapshotData
+          ? appLocalizationsOf(context).computingSnapshotData
+          : appLocalizationsOf(context).uploadingSnapshot,
+    ),
+    actions: [
+      if (onDismiss != null)
+        TextButton(
+          onPressed: onDismiss,
+          child: Text(appLocalizationsOf(context).cancel),
```
Comment (thiagocarvalhodev) [inline lib/components/create_snapshot_dialog.dart:99]: add a `,` after `cancel`

File: lib/blocs/create_snapshot/create_snapshot_cubit.dart [hunk 2]
```diff
@@ -182,9 +191,19 @@ class CreateSnapshotCubit extends Cubit<CreateSnapshotState> {
     // ignore: deprecated_export_use
     final dataBuffer = BytesBuilder(copy: false);
 
-    // Stream snapshot data to the temporal buffer
-    await for (final item in _newItemToBeCreated.getSnapshotData()) {
-      dataBuffer.add(item);
+    _snapshotDataStreamSubscription =
```
Comment (thiagocarvalhodev) [inline lib/blocs/create_snapshot/create_snapshot_cubit.dart:194]: nit: you can still use the `await for`.

```dart
await for (final item in _newItemToBeCreated.getSnapshotData()) {
      if(_wasSnapshotDataStreamSubscriptionCanceled) {
             throw Exception('canceled');
      }       
      dataBuffer.add(item);
}

 void cancelSnapshotCreation() async {
    _wasSnapshotDataStreamSubscriptionCanceled = true;
  }

**Review by thiagocarvalhodev — APPROVED** [review]
LGTM!

File: lib/blocs/create_snapshot/create_snapshot_cubit.dart [hunk 3]
```diff
@@ -289,6 +308,14 @@ class CreateSnapshotCubit extends Cubit<CreateSnapshotState> {
       emit(SnapshotUploadFailure(errorMessage: '$err'));
     }
   }
+
+  Future<void> cancelSnapshotCreation() async {
```
Comment (thiagocarvalhodev) [inline lib/blocs/create_snapshot/create_snapshot_cubit.dart:312]: We're missing a test for this method.

**Review by thiagocarvalhodev — CHANGES_REQUESTED** [review]

File: lib/blocs/create_snapshot/create_snapshot_cubit.dart (hunk 2 above)
Comment (matibat) [reply lib/blocs/create_snapshot/create_snapshot_cubit.dart:194]: I used listen here cuz it returns the subscription I can utilize to cancel the stream. I'm gonna try by simply throwing the way you suggested as the `await for` would consume the stream as well.

Comment (matibat) [reply lib/blocs/create_snapshot/create_snapshot_cubit.dart]: Much cleaner - a176c4c9ef66dda3ecffd9231b661fc9f85c56b7

File: lib/blocs/create_snapshot/create_snapshot_cubit.dart (hunk 3 above)
Comment (matibat) [reply lib/blocs/create_snapshot/create_snapshot_cubit.dart]: Something like this -> 8cef93d5a4c9db072e3047f3d4b43189ba2148e4

File: lib/components/create_snapshot_dialog.dart (hunk 1 above)
Comment (matibat) [reply lib/components/create_snapshot_dialog.dart]: 126adf839eee26898b8f1aea39fda169bbcced8a

**Review by thiagocarvalhodev — APPROVED** [review]
LGTM

File: lib/blocs/create_snapshot/create_snapshot_cubit.dart [hunk 4]
```diff
@@ -32,6 +33,8 @@ class CreateSnapshotCubit extends Cubit<CreateSnapshotState> {
   late Range _range;
   late int _currentHeight;
 
+  bool _wasSnapshotDataComputingCanceled = false;
```
[the remark this answers is not in the document]
Comment (matibat) [reply lib/blocs/create_snapshot/create_snapshot_cubit.dart:36]: https://ardrive.atlassian.net/browse/PE-2954

**Pull request merged**